File: cli/src/read.rs
use fancy_regex::Regex;
use fnv::{FnvBuildHasher, FnvHashMap};
use indexmap::IndexSet;
use rayon::prelude::*;
use serde_json::{from_str, Value};
use std::{
    fs::{read_dir, read_to_string, write, DirEntry},
    path::{Path, PathBuf},
};

/// Reads all Map .json files of input_path and parses them into .txt files in output_path.
/// # Parameters
/// * `input_path` - path to directory than contains .json files
/// * `output_path` - path to output directory
/// * `logging` - whether to log or not
/// * `log_string` - string to log
pub fn read_map(input_path: &Path, output_path: &Path, logging: bool, log_string: &str) {
    let re: Regex = Regex::new(r"^Map[0-9].*json$").unwrap();

    let files: Vec<DirEntry> = read_dir(input_path)
        .unwrap()
        .flatten()
        .filter(|entry: &DirEntry| {
            let filename: String = entry.file_name().into_string().unwrap();
            re.is_match(&filename).unwrap()
        })
        .collect();

    let maps_obj_map: FnvHashMap<String, Value> = files
        .iter()
        .map(|entry: &DirEntry| {
            (
                entry.file_name().into_string().unwrap(),
                from_str(&read_to_string(entry.path()).unwrap()).unwrap(),
            )
        })
        .collect();

    let mut lines: IndexSet<String, FnvBuildHasher> = IndexSet::default();
    let mut names_lines: IndexSet<String, FnvBuildHasher> = IndexSet::default();

    for (filename, obj) in maps_obj_map {
        if let Some(display_name) = obj["displayName"].as_str() {
            if !display_name.is_empty() {
                names_lines.insert(display_name.to_string());
            }
        }

        //Skipping first element in array as it is null
        for event in obj["events"].as_array().unwrap().iter().skip(1) {
            if !event["pages"].is_array() {
                continue;
            }

            for page in event["pages"].as_array().unwrap().iter() {
                let mut in_seq: bool = false;
                let mut line: Vec<String> = Vec::new();

                for list in page["list"].as_array().unwrap() {
                    //401 - dialogue lines
                    //102 - dialogue choices
                    //356 - system lines (special texts)
                    let code: u16 = list["code"].as_u64().unwrap() as u16;

                    for parameter_value in list["parameters"].as_array().unwrap() {
                        if code == 401 {
                            in_seq = true;

                            if parameter_value.is_string() {
                                let parameter: &str = parameter_value.as_str().unwrap();

                                if !parameter.is_empty() {
                                    line.push(parameter.to_string());
                                }
                            }
                        } else {
                            if in_seq {
                                lines.insert(line.join(r"\#"));
                                line.clear();
                                in_seq = false;
                            }

                            match code {
                                102 => {
                                    if parameter_value.is_array() {
                                        for param_value in parameter_value.as_array().unwrap() {
                                            if param_value.is_string() {
                                                let param: &str = param_value.as_str().unwrap();

                                                if !param.is_empty() {
                                                    lines.insert(param.to_string());
                                                }
                                            }
                                        }
                                    }
                                }

                                356 => {
                                    if parameter_value.is_string() {
                                        let parameter: &str = parameter_value.as_str().unwrap();

                                        if !parameter.is_empty() {
                                            lines.insert(parameter.to_string());
                                        }
                                    }
                                }

                                _ => {}
                            }
                        }
                    }
                }
            }
        }

        if logging {
            println!("{log_string} {filename}.");
        }
    }

    write(
        output_path.join("maps.txt"),
        lines.iter().cloned().collect::<Vec<String>>().join("\n"),
    )
    .unwrap();

    write(
        output_path.join("names.txt"),
        names_lines
            .iter()
            .cloned()
            .collect::<Vec<String>>()
            .join("\n"),
    )
    .unwrap();

    write(
        output_path.join("maps_trans.txt"),
        "\n".repeat(lines.len() - 1),
    )
    .unwrap();

    write(
        output_path.join("names_trans.txt"),
        "\n".repeat(names_lines.len() - 1),
    )
    .unwrap();
}

/// Reads all Other .json files of input_path and parses them into .txt files in output_path.
/// # Parameters
/// * `input_path` - path to directory than contains .json files
/// * `output_path` - path to output directory
/// * `logging` - whether to log or not
/// * `log_string` - string to log
pub fn read_other(input_path: &Path, output_path: &Path, logging: bool, log_string: &str) {
    let re: Regex = Regex::new(r"^(?!Map|Tilesets|Animations|States|System).*json$").unwrap();

    let files: Vec<DirEntry> = read_dir(input_path)
        .unwrap()
        .flatten()
        .filter(|entry: &DirEntry| {
            re.is_match(&entry.file_name().into_string().unwrap())
                .unwrap()
        })
        .collect();

    let obj_arr_map: FnvHashMap<String, Vec<Value>> = files
        .par_iter()
        .map(|entry: &DirEntry| {
            (
                entry.file_name().into_string().unwrap(),
                from_str(&read_to_string(entry.path()).unwrap()).unwrap(),
            )
        })
        .collect();

    for (filename, obj_arr) in obj_arr_map {
        let processed_filename: String = filename[0..filename.rfind('.').unwrap()].to_lowercase();
        let mut lines: IndexSet<String, FnvBuildHasher> = IndexSet::default();

        // Other files except CommonEvents.json and Troops.json have the structure that consists
        // of name, nickname, description and note
        if !filename.starts_with("Common") && !filename.starts_with("Troops") {
            for obj in obj_arr {
                if obj["name"].is_string() {
                    let name: &str = obj["name"].as_str().unwrap();

                    if !name.is_empty() {
                        lines.insert(name.to_string());
                    }
                }

                if obj["description"].is_string() {
                    let description: &str = obj["description"].as_str().unwrap();

                    if !description.is_empty() {
                        lines.insert(description.replace('\n', r"\#"));
                    }
                }

                if obj["nickname"].is_string() {
                    let nickname: &str = obj["nickname"].as_str().unwrap();

                    if !nickname.is_empty() {
                        lines.insert(nickname.to_string());
                    }
                }

                if obj["note"].is_string() {
                    let note: &str = obj["note"].as_str().unwrap();

                    if !note.is_empty() {
                        lines.insert(note.replace('\n', r"\#"));
                    }
                }
            }
        }
        //Other files have the structure somewhat similar to Maps.json files
        else {
            //Skipping first element in array as it is null
            for obj in obj_arr.iter().skip(1) {
                //CommonEvents doesn't have pages, so we can just check if it's Troops
                let pages_length: usize = if filename.starts_with("Troops") {
                    obj["pages"].as_array().unwrap().len()
                } else {
                    1
                };

                for i in 0..pages_length {
                    let list: &Value = if pages_length != 1 {
                        &obj["pages"][i]["list"]
                    } else {
                        &obj["list"]
                    };

                    if !list.is_array() {
                        continue;
                    }

                    let mut in_seq: bool = false;
                    let mut line: Vec<String> = Vec::with_capacity(256);

                    for list in list.as_array().unwrap() {
                        //401 - dialogue lines
                        //102 - dialogue choices
                        //356 - system lines (special texts)
                        //405 - credits lines
                        let code: u16 = list["code"].as_u64().unwrap() as u16;

                        for parameter_value in list["parameters"].as_array().unwrap() {
                            if code == 401 || code == 405 {
                                in_seq = true;

                                if parameter_value.is_string() {
                                    let parameter: &str = parameter_value.as_str().unwrap();

                                    if !parameter.is_empty() {
                                        line.push(parameter.to_string());
                                    }
                                }
                            } else {
                                if in_seq {
                                    lines.insert(line.join(r"\#"));
                                    line.clear();
                                    in_seq = false;
                                }

                                match code {
                                    102 => {
                                        if parameter_value.is_array() {
                                            for param_value in parameter_value.as_array().unwrap() {
                                                if param_value.is_string() {
                                                    let param: &str = param_value.as_str().unwrap();

                                                    if !param.is_empty() {
                                                        lines.insert(param.to_string());
                                                    }
                                                }
                                            }
                                        }
                                    }

                                    356 => {
                                        if parameter_value.is_string() {
                                            let parameter: &str = parameter_value.as_str().unwrap();

                                            if !parameter.is_empty() {
                                                lines.insert(parameter.to_string());
                                            }
                                        }
                                    }

                                    _ => {}
                                }
                            }
                        }
                    }
                }
            }
        }

        write(
            output_path.join(format!("{processed_filename}.txt")),
            lines.iter().cloned().collect::<Vec<String>>().join("\n"),
        )
        .unwrap();
        write(
            output_path.join(format!("{processed_filename}_trans.txt")),
            "\n".repeat(lines.len() - 1),
        )
        .unwrap();

        if logging {
            println!("{log_string} {filename}");
        }
    }
}

/// Reads System .json file of input_path and parses it into .txt file in output_path.
/// # Parameters
/// * `input_path` - path to directory than contains .json files
/// * `output_path` - path to output directory
/// * `logging` - whether to log or not
/// * `log_string` - string to log
pub fn read_system(input_path: &Path, output_path: &Path, logging: bool, log_string: &str) {
    let system_file: PathBuf = input_path.join("System.json");

    let obj: Value = from_str(&read_to_string(system_file).unwrap()).unwrap();

    let mut lines: IndexSet<String, FnvBuildHasher> = IndexSet::default();

    // Armor types names
    // Normally it's system strings, but might be needed for some purposes
    for string in obj["armorTypes"].as_array().unwrap() {
        let slice_ref: &str = string.as_str().unwrap();

        if !slice_ref.is_empty() {
            lines.insert(slice_ref.to_string());
        }
    }

    // Element types names
    // Normally it's system strings, but might be needed for some purposes
    for string in obj["elements"].as_array().unwrap() {
        let slice_ref: &str = string.as_str().unwrap();

        if !slice_ref.is_empty() {
            lines.insert(slice_ref.to_string());
        }
    }

    // Names of equipment slots
    for string in obj["equipTypes"].as_array().unwrap() {
        let slice_ref: &str = string.as_str().unwrap();

        if !slice_ref.is_empty() {
            lines.insert(slice_ref.to_string());
        }
    }

    // Game title, parsed just for fun
    // Translators may add something like "ELFISH TRANSLATION v1.0.0" to the title
    lines.insert(obj["gameTitle"].as_str().unwrap().to_string());

    // Names of battle options
    for string in obj["skillTypes"].as_array().unwrap() {
        let slice_ref: &str = string.as_str().unwrap();

        if !slice_ref.is_empty() {
            lines.insert(string.to_string());
        }
    }

    // Game terms vocabulary
    for (key, value) in obj["terms"].as_object().unwrap() {
        if key != "messages" {
            for string in value.as_array().unwrap() {
                if string.is_string() {
                    let slice_ref: &str = string.as_str().unwrap();

                    if !slice_ref.is_empty() {
                        lines.insert(slice_ref.to_string());
                    }
                }
            }
        } else {
            if !value.is_object() {
                continue;
            }

            for message_string in value.as_object().unwrap().values() {
                let slice_ref: &str = message_string.as_str().unwrap();

                if !slice_ref.is_empty() {
                    lines.insert(slice_ref.to_string());
                }
            }
        }
    }

    // Weapon types names
    // Normally it's system strings, but might be needed for some purposes
    for string in obj["weaponTypes"].as_array().unwrap() {
        let slice_ref: &str = string.as_str().unwrap();

        if !slice_ref.is_empty() {
            lines.insert(slice_ref.to_string());
        }
    }

    write(
        output_path.join("system.txt"),
        lines.iter().cloned().collect::<Vec<String>>().join("\n"),
    )
    .unwrap();
    write(
        output_path.join("system_trans.txt"),
        "\n".repeat(lines.len() - 1),
    )
    .unwrap();

    if logging {
        println!("{log_string} System.json.");
    }
}

// read_plugins is not implemented and will NEVER be, as plugins can differ from each other incredibly.
// Change plugins.js with your own hands.

File: cli/ruby-writer/read.ts
import { readdir } from "node:fs/promises";
import { OrderedSet } from "immutable";
import { load } from "@hyrious/marshal";
import { inflate } from "pako";

import { getValueBySymbolDesc } from "./symbol-utils";

const decoder = new TextDecoder();

const decode = (buffer: Uint8Array): string => {
    return decoder.decode(buffer);
};

function parseCode(code: number, parameter: (string | Uint8Array) | string[], gameType: string): string | void {
    switch (code) {
        case 401 || 405:
            if (parameter instanceof Uint8Array) {
                parameter = decode(parameter);
            }

            if (typeof parameter === "string" && parameter.length > 0) {
                switch (gameType) {
                    case "lisa":
                        const match = parameter.match(/^\\et\[[0-9]+\]/) ?? parameter.match(/^\\et\[[0-9]+\]/);

                        if (match) {
                            parameter = parameter.slice(match[0].length);
                        }
                        break;
                }

                return parameter;
            }
        case 102:
            if (parameter instanceof Uint8Array) {
                parameter = decode(parameter);
            }

            if (typeof parameter === "string" && parameter.length > 0) {
                return parameter;
            }
            break;
        case 356:
            if (parameter instanceof Uint8Array) {
                parameter = decode(parameter);
            }

            if (typeof parameter === "string" && parameter.length > 0) {
                return parameter as string;
            }
            break;
    }
}

function parseVariable(variable: string | Uint8Array, gameType: string): string | undefined {
    if (variable instanceof Uint8Array) {
        variable = decode(variable);
    }

    if (typeof variable === "string" && variable.length > 0) {
        const linesCount = variable.match(/\r?\n/g)?.length ?? 0;

        if (linesCount >= 1) {
            switch (gameType) {
                case "lisa":
                    variable = variable.replaceAll(/\r?\n/g, "\\#");

                    if (!variable.split("\\#").every((line) => /^<.*>\.?$/.test(line) || line.length === 0)) {
                        return variable;
                    } else {
                        return undefined;
                    }
                    break;
            }
        }
    }

    return variable;
}

/**
 * Reads all Map .rx/rv/rvdata2 files of inputPath and parses them into .txt files in outputPath.
 *
 * Based on the engine type, strings can be either UTF-8 encoded or binary.
 * This function handles both cases.
 * @param {string} inputPath - path to directory than contains .rx/rv/rvdata2 files
 * @param {string} outputPath - path to output directory
 * @param {boolean} logging - whether to log or not
 * @param {string} logString - string to log
 * @returns {Promise<void>}
 */
export async function readMap(
    inputPath: string,
    outputPath: string,
    logging: boolean,
    logString: string,
    gameType: string
): Promise<void> {
    const re = /^Map[0-9].*(rxdata|rvdata|rvdata2)$/;
    const files = (await readdir(inputPath)).filter((filename) => re.test(filename));

    const filesData: ArrayBuffer[] = await Promise.all(
        files.map((filename) => Bun.file(`${inputPath}/${filename}`).arrayBuffer())
    );

    const objMap = new Map(files.map((filename, i) => [filename, load(filesData[i]) as RubyObject]));

    const lines = OrderedSet().asMutable() as OrderedSet<string>;
    const namesLines = OrderedSet().asMutable() as OrderedSet<string>;

    for (const [filename, obj] of objMap) {
        let displayName: string | Uint8Array = getValueBySymbolDesc(obj, "@display_name");

        if (displayName instanceof Uint8Array) {
            displayName = decode(displayName);
        }

        if (typeof displayName === "string" && displayName.length > 0) {
            namesLines.add(displayName);
        }

        const events: object = getValueBySymbolDesc(obj, "@events");

        for (const event of Object.values(events || {})) {
            const pages: RubyObject[] = getValueBySymbolDesc(event, "@pages");
            if (!pages) {
                continue;
            }

            for (const page of pages) {
                let inSeq: boolean = false;
                const line: string[] = [];
                const list: RubyObject[] = getValueBySymbolDesc(page, "@list");

                if (!Array.isArray(list)) {
                    continue;
                }

                for (const item of list) {
                    //401 - dialogue lines
                    //102 - dialogue choices
                    //356 - system lines (special texts)
                    const code: number = getValueBySymbolDesc(item, "@code");

                    const parameters: (string | Uint8Array)[] | (string | Uint8Array)[][] = getValueBySymbolDesc(
                        item,
                        "@parameters"
                    );

                    for (const parameter of parameters as (string | Uint8Array)[]) {
                        if (code === 401) {
                            inSeq = true;

                            const parsed = parseCode(code, parameter, gameType);

                            if (parsed) {
                                line.push(parsed);
                            }
                        } else {
                            if (inSeq) {
                                lines.add(line.join("\\#"));
                                line.length = 0;
                                inSeq = false;
                            }

                            if (code === 102) {
                                if (Array.isArray(parameter)) {
                                    for (const param of parameter) {
                                        const parsed = parseCode(code, param, gameType);

                                        if (parsed) {
                                            lines.add(parsed);
                                        }
                                    }
                                }
                            } else if (code === 356) {
                                const parsed = parseCode(code, parameter, gameType);

                                if (parsed) {
                                    lines.add(parsed);
                                }
                            }
                        }
                    }
                }
            }
        }

        if (logging) {
            console.log(`${logString} ${filename}`);
        }
    }

    await Bun.write(`${outputPath}/maps.txt`, lines.join("\n"));
    await Bun.write(`${outputPath}/maps_trans.txt`, "\n".repeat(lines.size ? lines.size - 1 : 0));
    await Bun.write(`${outputPath}/names.txt`, namesLines.join("\n"));
    await Bun.write(`${outputPath}/names_trans.txt`, "\n".repeat(namesLines.size ? namesLines.size - 1 : 0));
}

/**
 * Reads all non-Map .rx/rv/rvdata2 files of inputPath (except System and Scripts) and parses them into .txt files in outputPath.
 *
 * Based on the engine type, strings can be either UTF-8 encoded or binary.
 * This function handles both cases.
 * @param {string} inputPath - path to directory than contains .rx/rv/rvdata2 files
 * @param {string} outputPath - path to output directory
 * @param {boolean} logging - whether to log or not
 * @param {string} logString - string to log
 * @returns {Promise<void>}
 */
export async function readOther(
    inputPath: string,
    outputPath: string,
    logging: boolean,
    logString: string,
    gameType: string
): Promise<void> {
    const re = /^(?!Map|Tilesets|Animations|States|System|Scripts|Areas).*(rxdata|rvdata|rvdata2)$/;
    const filenames = (await readdir(inputPath)).filter((filename) => re.test(filename));

    const filesData: ArrayBuffer[] = await Promise.all(
        filenames.map((filename) => Bun.file(`${inputPath}/${filename}`).arrayBuffer())
    );

    const objArrMap = new Map(
        // Slicing off the first element in array as it is null
        filenames.map((filename, i) => [filename, (load(filesData[i]) as RubyObject[]).slice(1)])
    );

    for (const [filename, objArr] of objArrMap) {
        const processedFilename = filename.toLowerCase().slice(0, filename.lastIndexOf("."));
        const lines = OrderedSet().asMutable() as OrderedSet<string>;

        // Other files except CommonEvents.json and Troops.json have the structure that consists
        // of name, nickname, description and note
        if (!filename.startsWith("Common") && !filename.startsWith("Troops")) {
            for (const obj of objArr) {
                const name: string | Uint8Array = getValueBySymbolDesc(obj, "@name");
                const nickname: string | Uint8Array = getValueBySymbolDesc(obj, "@nickname");
                const description: string | Uint8Array = getValueBySymbolDesc(obj, "@description");
                const note: string | Uint8Array = getValueBySymbolDesc(obj, "@note");

                for (const variable of [name, nickname, description, note]) {
                    const parsed = parseVariable(variable, gameType);

                    if (parsed) {
                        lines.add(parsed);
                    }
                }
            }
        }
        //Other files have the structure somewhat similar to Maps.json files
        else {
            for (const obj of objArr) {
                //CommonEvents doesn't have pages, so we can just check if it's Troops
                const pages: RubyObject[] = getValueBySymbolDesc(obj, "@pages");
                const pagesLength = filename.startsWith("Troops") ? pages.length : 1;

                for (let i = 0; i < pagesLength; i++) {
                    //If it's Troops, we'll iterate over the pages
                    //Otherwise we'll just iterate over the list
                    const list: RubyObject[] = filename.startsWith("Troops")
                        ? getValueBySymbolDesc(pages[i], "@list")
                        : getValueBySymbolDesc(obj, "@list");

                    if (!Array.isArray(list)) {
                        continue;
                    }

                    let inSeq: boolean = false;
                    const line: string[] = [];

                    for (const item of list) {
                        //401 - dialogue lines
                        //102 - dialogue choices
                        //356 - system lines (special texts)
                        //405 - credits lines
                        const code: number = getValueBySymbolDesc(item, "@code");
                        const parameters: (string | Uint8Array)[] | (string | Uint8Array)[][] = getValueBySymbolDesc(
                            item,
                            "@parameters"
                        );

                        for (const parameter of parameters as (string | Uint8Array)[]) {
                            if (code === 401 || code === 405) {
                                inSeq = true;

                                const parsed = parseCode(code, parameter, gameType);

                                if (parsed) {
                                    line.push(parsed);
                                }
                            } else {
                                if (inSeq) {
                                    lines.add(line.join("\\#"));
                                    line.length = 0;
                                    inSeq = false;
                                }

                                if (code === 102) {
                                    if (Array.isArray(parameter)) {
                                        for (const param of parameter) {
                                            const parsed = parseCode(code, param, gameType);

                                            if (parsed) {
                                                lines.add(parsed);
                                            }
                                        }
                                    }
                                } else if (code === 356) {
                                    const parsed = parseCode(code, parameter, gameType);

                                    if (parsed) {
                                        lines.add(parsed);
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }

        if (logging) {
            console.log(`${logString} ${filename}`);
        }

        await Bun.write(`${outputPath}/${processedFilename}.txt`, lines.join("\n"));
        await Bun.write(`${outputPath}/${processedFilename}_trans.txt`, "\n".repeat(lines.size ? lines.size - 1 : 0));
    }
}

/**
 * Reads System .rx/rv/rvdata2 file from inputFilePath and parses it into .txt file in outputPath.
 *
 * Based on the engine type, strings can be either UTF-8 encoded or binary.
 * This function handles both cases.
 * @param {string} systemFilePath - path to .rx/rv/rvdata2 file
 * @param {string} outputPath - path to output directory
 * @param {boolean} logging - whether to log or not
 * @param {string} logString - string to log
 * @returns {Promise<void>}
 */
export async function readSystem(
    systemFilePath: string,
    outputPath: string,
    logging: boolean,
    logString: string
): Promise<void> {
    const file = Bun.file(systemFilePath);
    const obj = load(await file.arrayBuffer()) as RubyObject;
    const ext = systemFilePath.slice(systemFilePath.lastIndexOf(".") + 1, systemFilePath.length);

    const lines = OrderedSet().asMutable() as OrderedSet<string>;

    const termsDesc = ext !== "rxdata" ? "@terms" : "@words";
    const symbolDescs = [
        "@elements",
        "@skill_types",
        "@weapon_types",
        "@armor_types",
        "@currency_unit",
        termsDesc,
        "@game_title",
    ];

    // Game damage elements names
    const elements = getValueBySymbolDesc(obj, symbolDescs[0]) as (string | Uint8Array)[] | undefined;
    // Game skill types names
    const skillTypes = getValueBySymbolDesc(obj, symbolDescs[1]) as (string | Uint8Array)[] | undefined;
    // Game weapon types names
    const weaponTypes = getValueBySymbolDesc(obj, symbolDescs[2]) as (string | Uint8Array)[] | undefined;
    // Game armor types names
    const armorTypes = getValueBySymbolDesc(obj, symbolDescs[3]) as (string | Uint8Array)[] | undefined;
    // Game currency unit name
    let currencyUnit = getValueBySymbolDesc(obj, symbolDescs[4]) as string | Uint8Array | undefined;
    // Game terms (vocabulary), called "words" in RPG Maker XP
    const terms = getValueBySymbolDesc(obj, symbolDescs[5]) as RubyObject;
    // Game title
    let gameTitle = getValueBySymbolDesc(obj, symbolDescs[6]) as Uint8Array | string;

    for (const arr of [elements, skillTypes, weaponTypes, armorTypes]) {
        if (!arr) {
            continue;
        }

        for (let string of arr) {
            if (string instanceof Uint8Array) {
                string = decode(string);
            }

            if (typeof string === "string" && string.length > 0) {
                lines.add(string);
            }
        }
    }

    if (currencyUnit instanceof Uint8Array) {
        currencyUnit = decode(currencyUnit);
    }

    if (typeof currencyUnit === "string" && currencyUnit.length > 0) {
        lines.add(currencyUnit);
    }

    const termsSymbols = Object.getOwnPropertySymbols(terms);

    for (let i = 0; i < termsSymbols.length; i++) {
        const value = terms[termsSymbols[i]] as Uint8Array | string[];

        if (value instanceof Uint8Array) {
            const decoded = decode(value);

            if (decoded.length > 0) {
                lines.add(decoded);
            }
            continue;
        }

        for (const string of value) {
            if (string.length > 0) {
                lines.add(string as string);
            }
        }
    }

    if (gameTitle instanceof Uint8Array) {
        gameTitle = decode(gameTitle);
    }

    if (typeof gameTitle === "string" && gameTitle.length > 0) {
        lines.add(gameTitle);
    }

    if (logging) {
        console.log(`${logString} ${file.name}`);
    }

    await Bun.write(`${outputPath}/system.txt`, lines.join("\n"));
    await Bun.write(`${outputPath}/system_trans.txt`, "\n".repeat(lines.size ? lines.size - 1 : 0));
}

/**
 * Reads Scripts .rx/rv/rvdata2 file from inputFilePath and parses it into .txt file in outputPath.
 * @param {string} scriptsFilePath - path to .rx/rv/rvdata2 file
 * @param {string} outputPath - path to output directory
 * @param {boolean} logging - whether to log or not
 * @param {string} logString - string to log
 * @returns {Promise<void>}
 */
export async function readScripts(
    scriptsFilePath: string,
    outputPath: string,
    logging: boolean,
    logString: string
): Promise<void> {
    const file = Bun.file(scriptsFilePath);
    const uintarrArr = load(await file.arrayBuffer(), { string: "binary" }) as Uint8Array[][];

    const fullCode = [];
    for (let i = 0; i < uintarrArr.length; i++) {
        // To convert Ruby code to string, it needs to be inflated
        // Then we decode our inflated string to the actual UTF-8 encoded string
        // And then replace \r\ns with \#s, which are out custom newline marks
        const codeString = decode(inflate(uintarrArr[i][2])).replaceAll(/\r?\n/g, "\\#");
        fullCode.push(codeString);
    }

    if (logging) {
        console.log(`${logString} ${file.name}.`);
    }

    const joinedCode = fullCode.join("\n");
    await Bun.write(`${outputPath}/scripts.txt`, joinedCode);
    await Bun.write(`${outputPath}/scripts_trans.txt`, joinedCode);
}
